### ISSUE TITLE
fetchAllUsers() does not return expected users

### ISSUE THREAD

**Issue by 2FWAH** [carpedm20/fbchat#282]
Hello,

When using client.fetchAllUsers() the returned list does not contain all users whom I'm currently chatting with.
In addition some contacts that are friends but I don't talk with are in the list...

Do any of you have this problem?

**Comment by madsmtm** [comment]
Hmm, I think the name is taken from the URL, `"https://www.facebook.com/chat/user_info_all"`, though it may not be totally correct or perhaps outdated. My initial guess would be that some of the accounts are Messenger only accounts? Though that works for me, too.

I compared the result on my account to the list I got from `https://www.facebook.com/{ME}/friends`.

A few deleted accounts and a single memorial page showed up on the second list, and not on the first (These you would "expect". They're your friends, but not people you can chat with).

And a few showed up on the second list, and not on the first (again, mostly like you'd expect):
- A few account that I'm not friends with (testing accounts, other people that I've chatted with but aren't friends with)
- "Messenger only"-accounts
- A single person, whom I am friends with. Only when I searched for their name could I verify that they wre actually my Facebook friend, so I don't know what's going on there.

Also worth noting: I had a test account removed a while ago, which I was friends with, that showed up in neither of those two lists.

Tl;dr: Could confirm the parts that (most) Facebook friends are returned, but every person that I'm chatting with is also returned. Of course, keep in mind that pages and groups are not a part of this. Basically, I think it's a list of people you *can* chat with, not just *are* chatting with.

**Comment by 2FWAH** [comment]
Thanks for your detailed answer :)

My initial message (especially the "expected") was not really clear.
To clarify: IMO it would be logical to find _at least_ all people whom I'm currently chatting with (even friend or not, basically all people in all conversations)

Like you I have in the list returned by fetchAllUsers() Facebook friends I have never spoken to over Messenger/FB ("people you _can_ chat with"). IMO it's not an issue.

On my side it looks like that people whom I chatting with and not in client.fetchAllUsers() are not my friend... except for two of them (in the list but not in my friend list).
It doesn't seem to be related to the date of the thread.

Maybe fetchAllUsers() could be a wrapper function that will parse all threads and enumerate all users?
This would be more reliable?

**Comment by madsmtm** [comment]
Just to make sure I understand, you have a group-thread, where there's a few participants/members of that thread, that you're not friends with, or have chatted with, and they are not showing up in `fetchAllUsers`?

**Comment by 2FWAH** [comment]
On my side I got:
- one ONE_TO_ONE conversation with a non-friend is returned by `fetchAllUsers`
- another ONE_TO_ONE conversation with a non-friend (basically the same case) is _not_ returned by `fetchAllUsers`.
- `fetchAllUsers` returns users who are my friends with whom I have never spoken before.

To answer your specific question, `fetchAllUsers` seems to not return users in group-thread if I'm not friend with them.

`fetchAllUsers` seems to return all possible ONE_TO_ONE conversations : accepted conversations (non-friend people), conversations with your friends and people that your are able to chat with (friends).
But there are some exceptions (like accepted conversations not in the list), that is the problem...

It's possible that the endpoint is not used anymore by Messenger, thus the results are not accurate.

I did not success to find when a conversation is returned or not by `fetchAllUsers`.

I wrote a function to parse all users from all threads, should I PR it? (as `fetchAllUsers`?)

**Comment by madsmtm** [comment]
What do you mean by a function to "parse all users from all threads"? But yes, if you have something that works like `fetchAllUsers`, except consistently, then feel free to submit it!

**Comment by 2FWAH** [comment]
Hello,

Sorry for the delay.
I'm thinking about looping on all threads, and for each thread fetching all users in the thread.

I will PR a POC soon :)
